### PR TITLE
Read configuration from terraform.tfvars

### PR DESCRIPTION
**Pull request by brikis98**
This PR changes Terragrunt so that, by default, it reads its configuration from `terraform.tfvars` instead of `.terragrunt`. The reasons for this change are:

1. We piggyback on top of a file format Terraform users are already used to.
1. We reduce the number of files you need. This is especially valuable with [remote Terraform configurations](https://github.com/gruntwork-io/terragrunt#remote-terraform-configurations), as it means you only need a single `terraform.tfvars` file to define what code to download and provide the variables for that code.

For backwards compatibility, I’ve made sure that Terragrunt can still use `.terragrunt` files as before, though it will now log a deprecation warning. 

Note, this is a long PR, but just about all of the important stuff is in two files: `README.md` and `config.go`. Please focus your code review efforts there. The rest of the changes are:

1. Test code to check the new functionality.
1. Test code to check backwards compatibility with the old `.terragrunt` configuration format.
1. Updating lots of comments, docs, and test cases to not explicitly say `.terragrunt`. We had that copy/pasted in too many places.
1. An unrelated change regarding filepaths. We need to use `util.JoinPath` for paths to work correctly on all platforms. I should've done this fix as a separate PR, but got impatient.

### CODE REVIEW

**Comment by brikis98** [comment]
@josh-padnick This is a pretty significant change, so please take a look when you have a chance.

**Comment by brikis98** [comment]
@josh-padnick Reminder.

**Comment by josh-padnick** [comment]
@brikis98 My apologies for the delay on this. Reviewing now...

File: README.md [hunk 1]
```diff
@@ -60,34 +60,37 @@ a simple, free locking mechanism, and enforcing best practices around CLI usage
...
+    backend = "s3"
+    config {
+      encrypt = "true"
+      bucket = "my-bucket"
+      key = "terraform.tfstate"
+      region = "us-east-1"
+    }
   }
 }
 ```
```
Comment (brikis98) [inline README.md]: > What happens when Terraform itself encounters these directives in a terraform.tfvars? Does it just ignore them?

Yup, it just ignores them.

> I believe we previously implemented various interpolation functionality. Do test validate that still works as expected?

Yup, the automated tests check a lot of interpolation functionality in `.tfvars` files.

> We should still explicitly clarify the above points.

Good idea. Added: https://github.com/gruntwork-io/terragrunt/pull/122/commits/c550d8f47243ee4854bd0c1edce5470d1df6e772

File: README.md [hunk 2]
```diff
@@ -264,81 +271,86 @@ remote_state = {
   
   > Note: Terragrunt will use the value provided in the `profile` key to configure the AWS SDK when using the S3 backend.
 
-## Managing multiple .terragrunt files
+## Managing multiple Terraform configurations
 
-With Terraform, it can be a good idea to store your templates in separate folders (and therefore, separate state files)
-to provide isolation between different environments,such as stage and prod, and different components, such as a 
+With Terraform, it can be a good idea to store your configurations in separate folders (and therefore, separate state 
+files) to provide isolation between different environments,such as stage and prod, and different components, such as a
```
Comment (brikis98) [inline README.md]: Added: https://github.com/gruntwork-io/terragrunt/pull/122/commits/c550d8f47243ee4854bd0c1edce5470d1df6e772

File: README.md [hunk 3]
```diff
@@ -351,106 +363,117 @@ The solution is to use the following features of Terragrunt:
...
 Each of these items is discussed next.
 
 ### find_in_parent_folders helper
 
-Terragrunt supports the use of a few helper functions using the same syntax as Terraform: `${some_function()}`. One of
-the supported helper functions is `find_in_parent_folders()`, which returns the path to the first `.terragrunt` file it
-finds in the parent folders above the current `.terragrunt` file. 
+Terragrunt supports the use of a few helper functions in the `terraform.tfvars` file using the same syntax as 
+Terraform: `${some_function()}`. Note that these helper functions are *only* evaluated by Terragrunt and not Terraform,
+so they won't work outside of the `terragrunt = { ... }` block in the `terraform.tfvars` file.
```
Comment (brikis98) [inline README.md]: Terraform ignores it.

File: README.md [hunk 4]
```diff
@@ -483,68 +508,83 @@ parent, but the `remote_state` settings of the child.
 
 ### Environment variables replacement
 
-You can read in environment variables in your `.terragrunt` file using the `get_env()` helper function:
+You can read in environment variables in your `terraform.tfvars` file using the `get_env()` helper function:
```
Comment (brikis98) [inline README.md]: Fixed: https://github.com/gruntwork-io/terragrunt/pull/122/commits/b9a004184974037dccd2871ddb48d759270060e8

File: README.md [hunk 5]
```diff
@@ -483,68 +508,83 @@ parent, but the `remote_state` settings of the child.
...
+      bucket = "value-bucket"
+      key = "/foo/bar/terraform.tfstate"
+      region = "us-west-2"
+    }
   }
 }
 ```
 
-Terraform itself also supports loading variables via the environment. Is possible to use the same variables by correctly using the terrraform prefix `TF_VAR_`.
+Terraform itself also supports loading variables via the environment. Is possible to use the same variables by
```
Comment (brikis98) [inline README.md]: Fixed: https://github.com/gruntwork-io/terragrunt/pull/122/commits/b9a004184974037dccd2871ddb48d759270060e8

File: README.md [hunk 6]
```diff
@@ -728,79 +772,94 @@ environments should be exposed as an input variable. For example, the frontend-a
...
-    instance_count = 10
-    instance_type = "m2.large"
-    ```
-1. `.terragrunt`: This is the typical Terragrunt configuration file, which specifies how to manage remote state and 
-   locking. 
+instance_count = 3
+instance_type = "t2.micro"
+```
+
+And `prod/frontend-app.tfvars` may look like this:
```
Comment (brikis98) [inline README.md]: You're right, the structure of the docs was confusing. It's a leftover from when the `source` stuff was defined in a separate file (`.terragrunt`). I've re-ordered it for clarity here: https://github.com/gruntwork-io/terragrunt/pull/122/commits/56c04585074fe4df78f7e7ceba29416d66fa7854

File: README.md [hunk 7]
```diff
@@ -728,79 +772,94 @@ environments should be exposed as an input variable. For example, the frontend-a
...
-terraform {
-  source = "git::git@github.com:foo/bar.git//frontend-app?ref=v0.0.1"
+terragrunt = {
+  include {
+    path = "${find_in_parent_folders()}"
+  }
+  
+  terraform {
+    source = "git::git@github.com:foo/bar.git//frontend-app?ref=v0.0.1"
+  }
```
Comment (brikis98) [inline README.md]: I re-ordered the docs for clarity: https://github.com/gruntwork-io/terragrunt/pull/122/commits/56c04585074fe4df78f7e7ceba29416d66fa7854

File: README.md [hunk 8]
```diff
@@ -728,79 +772,94 @@ environments should be exposed as an input variable. For example, the frontend-a
...
 sources](https://www.terraform.io/docs/modules/sources.html).)
 
-When you run Terragrunt and it finds a `terraform` block in your configuration, Terragrunt will:
+When you run Terragrunt and it finds a `terraform` block, it will:
  
 1. Download the configurations specified via the `source` parameter into a temporary folder. This is done using the
    [terraform init command](https://www.terraform.io/docs/commands/init.html), so the `source` parameter supports the
    exact same syntax as the [module source](https://www.terraform.io/docs/modules/sources.html) parameter, including
    local file paths, Git URLs, and Git URLs with `ref` parameters (useful for checking out a specific tag, commit, or
    branch of Git repo). Terragrunt will download all the code in the repo (i.e. the part before the double-slash `//`)
```
Comment (brikis98) [inline README.md]: 1. Tried to clarify it a bit here: https://github.com/gruntwork-io/terragrunt/pull/122/commits/6c9f37be58a90d3b3515c85697d5b45d2f754ab2
1. It looks backwards compatible.

File: README.md [hunk 9]
```diff
@@ -728,79 +772,94 @@ environments should be exposed as an input variable. For example, the frontend-a
...
+environment-specific values in the `.tfvars` files, which is about as close to the "essential complexity" of the 
+problem as you can get.
 
-Just as importantly, since the Terraform code is now defined in a single repo, you can version it (e.g. using Git
+Just as importantly, since the Terraform code is now defined in a single repo, you can version it (e.g., using Git
 tags and referencing them using the `ref` parameter in the `source` URL), and promote a single, immutable version 
-through each environment (e.g. qa -> stage -> prod). This idea is inspired by Kief Morris' blog post [Using Pipelines 
+through each environment (e.g., qa -> stage -> prod). This idea is inspired by Kief Morris' blog post [Using Pipelines 
 to Manage Environments with Infrastructure as 
 Code](https://medium.com/@kief/https-medium-com-kief-using-pipelines-to-manage-environments-with-infrastructure-as-code-b37285a1cbf5).
```
Comment (brikis98) [inline README.md]: Clarified this a bit: https://github.com/gruntwork-io/terragrunt/pull/122/commits/ea0f8cac4f1ac64f7cb41e151da8be8894a14bea

File: README.md [hunk 10]
```diff
@@ -838,6 +897,13 @@ prefix `--terragrunt-`. The currently available options are:
 * `--terragrunt-source-update`: Delete the contents of the temporary folder before downloading Terraform source code
   into it.
 
+## The old .terragrunt configuration file
+
+Older versions of Terragrunt used to read their configuration from a `.terragrunt` file instead of `terraform.tfvars`.
+This file format is now **deprecated**, although for backwards compatibility, Terragrunt will continue to support it
+for the near future. The configuration in a `.terragrunt` file is identical to that of the `terraform.tfvars` file 
+except it does *not* need to be wrapped with a `terragrunt = { ... }` block.
+
```
Comment (brikis98) [inline README.md]: Added explicit docs on migrating from `.terragrunt`: https://github.com/gruntwork-io/terragrunt/pull/122/commits/e6f556925af497a32318688f28b94162c58b2074

Comment (brikis98) [inline README.md]: Agreed these docs are getting longer. Clean up for that definitely belongs in a separate PR.

File: README.md [hunk 11]
```diff
@@ -919,14 +985,3 @@ See `circle.yml` and `_ci/build-and-push-release-asset.sh` for details.
...
-## TODO
-
-* Add a check that modules have been downloaded using `terraform get`.
-* Add a check that all local changes have been committed before running `terraform apply`.
-* Consider implementing alternative locking mechanisms, such as using Git instead of DynamoDB.
-* Consider embedding the Terraform Go code within Terragrunt instead of calling out to it.
-* Add a `show-lock` command.
-* Add a command to automatically set up best-practices remote state storage in a versioned, encrypted, S3 bucket.
-* Add a command to list the different versions of state available in a versioned S3 bucket and to diff any two state
-  files.
```
Comment (brikis98) [inline README.md]: Added migration steps here: https://github.com/gruntwork-io/terragrunt/pull/122/commits/e6f556925af497a32318688f28b94162c58b2074

File: config/config.go [hunk 12]
```diff
@@ -62,6 +71,87 @@ func (conf *TerraformConfig) String() string {
...
+// Return the default path to use for the Terragrunt configuration file. The reason this is a method rather than a
+// constant is that older versions of Terragrunt stored configuration in a different file. This method returns the
+// path to the old configuration format if such a file exists and the new format otherwise.
+func DefaultConfigPath(workingDir string) string {
+	path := util.JoinPath(workingDir, OldTerragruntConfigPath)
+	if util.FileExists(path) {
+		return path
+	}
+	return util.JoinPath(workingDir, DefaultTerragruntConfigPath)
+}
```
Comment (brikis98) [inline config/config.go]: Added: https://github.com/gruntwork-io/terragrunt/pull/122/commits/e6f556925af497a32318688f28b94162c58b2074

**Comment by brikis98** [comment]
Merging now.
